### PR TITLE
[jit] Tree based Iterator infrastructure: for in range/list/tensor/zip/enumerate

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22006 [jit] add for in dict support
* #21990 [jit] add for in string support
* #21985 [jit] support iteration tuple unpacking
* #21950 [jit] refactor all for in range/tensor tests to be together with other for loop tests
* **#21801 [jit] Tree based Iterator infrastructure: for in range/list/tensor/zip/enumerate**
* #21949 [jit] more complete tuple assignments

Summary:

This makes the looping infrastructure on different types of iterators be tree based. SimpleValue(List, Dict, Tensor, etc) will be leaves of the tree, they know how to calculate the loop information using `len()` and `get_elem()`, while builtins like `zip` or `enumerate` become non-leaf of the tree. 

When a users writes:
```
for (i, (j, k), l)  in zip(a, enumerate(b), range(0, 100)):
```   
    
We generate:

```
IterableTree {
    root = ((), ((), ()), ())
    bases = {a, range(0, math.inf, 1), b, range(0, 100)};
}

for _i in range(0, prim::min(len(a), len(b), len(range(0, 100)))):
    b_0_i = a[_i]    
    b_1_i = range(0, math.inf, 1)[_i]    
    b_2_i = b[_i]    
    b_3_i = range(0, 100)[_i]
    i, (j, k), l = (b_0_i, (b_1_i, b_2_i), b_3_i)
```

Follow ups:
* Iteration Tuple unpacking
* for in Dict
* for in String
* reverse builtin

Differential Revision: [D15948545](https://our.internmc.facebook.com/intern/diff/D15948545)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -3279,96 +3279,6 @@ def stuff3(x):
             return torch.ones(x), x
         self.checkScript(stuff3, ([3, 2],))
 
-    def test_for_in_tensors(self):
```
Comment (zdevito) [inline test/test_jit.py]: This move makes it really hard to see what tests were added for this PR. Can you tell me the test names that were added/modified or split this movement into a separate PR?

File: test/test_jit.py [hunk 2]
```diff
@@ -9890,14 +9719,308 @@ def return_stmt(x):
...
             def range_no_arg(x):
                 for _ in range():
                     x += 1
                 return x
 
+    def test_for_in_enumerate(self):
+        def fn(x):
+            # type: (List[int]) -> int
+            sum = 0
+            for (i, v) in enumerate(x):
```
Comment (zdevito) [inline test/test_jit.py]: `for i, v in enumerate(x)` is valid syntax, and will need to work. Changes for it can be in a separate PR.

File: torch/csrc/jit/script/sugared_value.h [hunk 3]
```diff
@@ -404,6 +416,64 @@ struct TORCH_API IsInstanceValue : SugaredValue {
...
+// Specialized Tree structure to matched against for special handling 
+// of builtin functions iterables expressions like zip(), enumerate(), etc.
+// zip and enumerate can be modeled as a tree of SimpleValue/RangeValue:
+//    zip(x, y) ->  (x, y) with tuple assignment to each loop target
+//    enumerate(x) -> (range(0, math.inf, 1), x)
+// So a complicated expression like zip(a, enumerate(b), range(0, 100)) will be:
+// (a, (range(0, math.inf, 1), b), range(0, 100))
+// We use those base iterables to fill in the loop information like max_trip_count
+// and set the value table for loop targets
+struct TORCH_API IterableTree : SugaredValue {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: Why does this not define get_elem and length?

File: torch/csrc/jit/script/sugared_value.h [hunk 4]
```diff
@@ -94,6 +94,15 @@ struct TORCH_API SugaredValue
     throw ErrorReport(loc) << "cannot call a " << kind();
   }
 
+  // return length of this thing, if not then it can't be iterated.
+  virtual Value* len(const SourceRange& loc, Function& m) {
+    throw ErrorReport(loc) << "cannot get the length of value " << kind();
+  }
+  // expression for ith elemement for iterable value
+  virtual Value* get_elem(const SourceRange&loc, Function& m, Value* i) {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: this should be call `getelem` to match Python's `__getelem__`

File: torch/csrc/jit/script/sugared_value.cpp [hunk 5]
```diff
@@ -49,6 +49,12 @@ std::shared_ptr<SugaredValue> BuiltinFunction::call(
     at::ArrayRef<NamedValue> inputs,
     at::ArrayRef<NamedValue> attributes,
     size_t n_binders) {
+
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: Range does not make sense as a BuiltinFunction -- it completely ignores the entire ::call infrastructure BuiltinFunction uses. It should be its own thing.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 6]
```diff
@@ -236,6 +242,122 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
...
+  // List, Tuple, Tensor, fill in missing information desugaring
+  Value* val = getValue();
+  TypePtr val_type = val->type();
+  Graph& g = *m.graph();
+  if (val_type->cast<ListType>()) {
+    return g.insert(aten::len, {val}, {}, loc);
+  } else if (val_type->isSubtypeOf(TensorType::get())) {
+    Value* outermost_dim_index = g.insertConstant(0, IntType::get(), loc);
+    // zero-dim tensor error handling
+    Value* num_dim = g.insert(aten::dim, {val}, {}, loc);
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: I am not a fan of using so much graph manipulation to write code. I'd prefer this just was a call to a builtin that got the length of the tensor or errored on zero dim.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 7]
```diff
@@ -236,6 +242,122 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
...
+      g.insert(prim::RaiseException,
+          {std::string("iteration over a 0-d tensor!")}, {}, loc);
+    }
+
+    // fill in max_trip_count_val
+    Value* sizes_tuple = g.insert(aten::size, {val}, {}, loc);
+    return g.insert(aten::select, {sizes_tuple, outermost_dim_index}, {}, loc);
+  } else {
+    throw ErrorReport(loc)
+      << "Value type " << val_type->str() << " does not length information";
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: This should say  `val_type->python_str()` "is not iterable".

File: torch/csrc/jit/script/compiler.cpp [hunk 8]
```diff
@@ -1288,8 +1292,8 @@ struct to_ir {
 
       // current_element_assigner uses an induction variable
       // to set a current element
-      if (current_element_assigner) {
-        current_element_assigner(trip_count, environment_stack);
+      if (iter_val != nullptr && targets) {
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: The comment above is now nonsense, and I can't follow what iter_val or targets are in this context.

File: torch/csrc/jit/script/compiler.cpp [hunk 9]
```diff
@@ -1298,246 +1302,92 @@ struct to_ir {
...
-      end_val->node()->setSourceRange(range);
-    } else if (args.size() == 2 || args.size() == 3) {
-      start_val = argVals[0];
-      end_val = argVals[1];
-      if (args.size() == 3) {
-        step_val = argVals[2];
-      } else {
-        step_val = end_val->owningGraph()->insertConstant(1);
-        step_val->node()->setSourceRange(range);
+  // recursively assign the targets base on the iterable tree structure
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: I do not think it is a good idea to special case the iterable_tree here. There are many places in the compiler that already do tuple assignment, e.g.
`a, (b, c)` = foo()`

We should be reusing that code for assign the loop lhs to whatever the result of getitem is.  The IterableTree can simply return a tree from that.

File: torch/csrc/jit/script/compiler.cpp [hunk 10]
```diff
@@ -1298,246 +1302,92 @@ struct to_ir {
...
+    // We will get IterableTree for builtinFunctions zip() and enumerate(),
+    // RangeValue for range(), and SimpleValue for types like List, Tensor, Dict.
+    auto range_val = std::dynamic_pointer_cast<RangeValue>(sv);
     auto siv = std::dynamic_pointer_cast<SimpleValue>(sv);
+    auto iterable_tree = std::dynamic_pointer_cast<IterableTree>(sv);
 
-    // for-in lists
-    if (siv && siv->getValue()->type()->kind() == TypeKind::ListType) {
-      emitForInListLoop(stmt, siv);
+    if ((siv && (siv->getValue()->type()->kind() == TypeKind::ListType
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: These branches for different iteration kinds should be avoided. We need to fit all iteraiton into the len/getitem abstraction.

File: torch/csrc/jit/script/compiler.cpp [hunk 11]
```diff
@@ -2377,6 +2229,60 @@ struct to_ir {
...
+  // since we could get the loop information from len() and get_item().
+  // For IterableValue like zip(), enumerate(), we can model them as a combination of
+  // leaves, and we emit a IterableTree value to record the tree information
+  SugaredValuePtr emitIterableTree(
+      SourceRange& loc,
+      const List<Expr>& inputs,
+      const std::shared_ptr<IterableValue>& iterable) {
+      std::shared_ptr<IterableTree> iterable_tree = nullptr;
+      size_t input_size = inputs.size();
+      if (iterable->symbol_ == prim::enumerate) {
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Could this be done on construction of the sugared value, to avoid having hold a prim::enumerate symbol to direct what to do later?

File: torch/csrc/jit/script/sugared_value.h [hunk 12]
```diff
@@ -404,6 +416,64 @@ struct TORCH_API IsInstanceValue : SugaredValue {
...
+    return "range";
+  }
+  Value* len(const SourceRange& loc, Function& m) override;
+  Value* get_elem(const SourceRange&loc, Function& m, Value* i) override;
+
+  private:
+    Value* start_;
+    Value* end_;
+    Value* step_;
+    bool is_simple_;
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: What does this mean?

**Review by zdevito — COMMENTED** [review]
This is a good start. Next steps:

1. let's try to get everything into the len/getitem abstraction, include the IterableTree. 
2. the destructuring assignment to assign to the lists iteration variables should use the same code as normal assignments, so it handles tuple restructuring correctly.

**Comment by suo** [comment]
wat

File: torch/csrc/jit/script/sugared_value.h (hunk 12 above)
Comment (wanchaol) [reply torch/csrc/jit/script/sugared_value.h]: `is_simple_` means we the input args of `range` only contains `end`, and this serves as a flag in `len` and `getelem` to not insert unnecessary builtin nodes.

File: torch/csrc/jit/script/compiler.cpp (hunk 9 above)
Comment (wanchaol) [reply torch/csrc/jit/script/compiler.cpp]: does that mean we should make the `getitem` to return `SugaredValue` instead of `Value*`?

File: torch/csrc/jit/script/compiler.cpp (hunk 11 above)
Comment (wanchaol) [reply torch/csrc/jit/script/compiler.cpp]: What I observed from the `emitApplyExpr` is that, it will first do `emitSugaredExpr(apply.callee(), 1)` without passing inputs in, and the `builtins` will emit corresponding sugared values, I could not do it there since it's missing the inputs, that's where I need the singleton `IterableValue` sugared value to record the symbol and do the later tree construction. I could instead do a whole special casing on the `IterableTree` when `emitApplyExpr`, but that makes it separate from the `builtin` registration table mechanism that we already using..

File: torch/csrc/jit/script/sugared_value.cpp (hunk 6 above)
Comment (wanchaol) [reply torch/csrc/jit/script/sugared_value.cpp]: Most of the comments are addressed. For this one, I just followed the old implementation on tensor looping. I thought this a bit, agree with you that we should probably register a builtin prim operator to hide the graph manipulations. Do we have the flexibility of adding prim operators for those graph manipulations?

File: test/test_jit.py [hunk 13]
```diff
@@ -3286,7 +3286,7 @@ def test_sizes(x):
             test_sizes(torch.tensor(1))
 
     def test_for_in_tensors_fail_scalar(self):
-        with self.assertRaisesRegex(RuntimeError, "cannot be used as a tuple"):
+        with self.assertRaisesRegex(RuntimeError, "cannot get length of the value type float"):
```
Comment (zdevito) [inline test/test_jit.py]: These error messages are not very good. It should say something about not being iterable. What is the python error message in this case? Can we match it?

File: torch/csrc/jit/script/parser.cpp [hunk 14]
```diff
@@ -256,7 +256,10 @@ struct ParserImpl {
       }
 
       if (kind == TK_FOR) {
-        auto target = parseIdent();
+        // TK_FOR targets should only parse exprs prec greater than 4, which only
+        // includes subset of Exprs that suppose to be on the LHS according to the
+        // python grammer https://docs.python.org/3/reference/grammar.html
+        auto target = parseExp(4);
```
Comment (zdevito) [inline torch/csrc/jit/script/parser.cpp]: I think this is meant to call parseLhsExp, otherwise it is unused.

File: torch/csrc/jit/script/parser.cpp [hunk 15]
```diff
@@ -329,6 +332,12 @@ struct ParserImpl {
     });
   }
 
+  // parse LHS acceptable exprs, which only includes subset of Exprs that prec is
+  // greater than 4 according to the python grammer
+  Expr parseLhsExp() {
```
Comment (zdevito) [inline torch/csrc/jit/script/parser.cpp]: nit: parseLHSExp

File: torch/csrc/jit/script/sugared_value.cpp [hunk 16]
```diff
@@ -259,6 +259,135 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
...
+      start_ = g.insertConstant(0, nullptr, loc);
+      step_ = g.insertConstant(1, nullptr, loc);
+      // range() call only contains end, easier to calculate len() and getitem()
+      is_simple_ = true;
+    } else if (inputs.size() <= 3) {
+      start_ = inputs[0];
+      end_ = inputs[1];
+      if (inputs.size() == 3) {
+        step_ = inputs[2];
+        // error handling when step_val = 0 during runtime
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: Can we make __range_length just do this error handling? It is much more inline with how we already do error handling everywhere else.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 17]
```diff
@@ -259,6 +259,135 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
...
+  // and then calculate the minimum length of all the base iterables to be max_trip_count_val
+  Graph& g = *m.graph();
+  std::vector<SugaredValuePtr> base_iters = get_base_iterables();
+  std::vector<Value*> lengths;
+  lengths.reserve(base_iters.size());
+
+  for (const SugaredValuePtr& base_iter: base_iters) {
+    lengths.emplace_back(base_iter->len(loc, m));
+  }
+  Node* list_node = g.insertNode(g.create(prim::ListConstruct, 1)->setSourceRange(loc));
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.cpp]: use node->createList please

File: torch/csrc/jit/script/sugared_value.h [hunk 18]
```diff
@@ -94,6 +94,15 @@ struct TORCH_API SugaredValue
     throw ErrorReport(loc) << "cannot call a " << kind();
   }
 
+  // return length of this thing, if not then it can't be iterated.
+  virtual Value* len(const SourceRange& loc, Function& m) {
```
Comment (zdevito) [inline torch/csrc/jit/script/sugared_value.h]: Can we make these error message match what python says in these circumstances? Since these are only used in iteration, it should be talking about how something is/isn't iterable.

File: torch/csrc/jit/script/sugared_value.h (hunk 12 above)
Comment (zdevito) [reply torch/csrc/jit/script/sugared_value.h]: maybe use `has_start_and_step_` then?

**Review by zdevito — APPROVED** [review]
Looks good, just minor issues

File: torch/csrc/jit/script/compiler.cpp [hunk 19]
```diff
@@ -2420,6 +2229,65 @@ struct to_ir {
...
+      const std::shared_ptr<IterableValue>& iterable) {
+      std::shared_ptr<IterableTree> iterable_tree = nullptr;
+      size_t input_size = inputs.size();
+
+      // Handling different iterable values
+      if (iterable->symbol_ == prim::range) {
+        std::vector<Value*> input_vals = getValues(inputs, /*maybe_unpack=*/true);
+        return std::make_shared<RangeValue>(loc, method, input_vals);
+      } else if (iterable->symbol_ == prim::enumerate) {
+        // enumerate(x) can be rewrite as subtrees: (range(0, math.inf), x)
```
Comment (Krovatkin) [inline torch/csrc/jit/script/compiler.cpp]: maybe we could expand your comment into `IterableTree(RangeValue(0, math.inf), SimpleValue(x)`

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/script/sugared_value.cpp [hunk 20]
```diff
@@ -259,6 +259,135 @@ std::shared_ptr<SugaredValue> SimpleValue::call(
...
+          << "range expected at most 3 arguments, got " << inputs.size();
+    }
+}
+
+Value* RangeValue::len(const SourceRange& loc, Function& m) {
+  if (is_simple_) {
+    return end_;
+  } else{
+    Graph& g = *m.graph();
+    return g.insert(aten::__range_length, {start_, end_, step_}, {}, loc);
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/csrc/jit/script/sugared_value.cpp]: Yes we don't want to insert the node when it's not necessary to do so. I will update with more complete comments

File: torch/csrc/jit/register_prim_ops.cpp [hunk 21]
```diff
@@ -2063,6 +2063,22 @@ RegisterOperators reg2({
...
     DEFINE_BINARY_OP(prim::max, a > b ? a : b),
+
+    Operator(
+        "prim::min(int[] x) -> int",
+        [](Stack& stack) {
+          c10::ListPtr<int64_t> int_list = pop(stack).toIntList();
+          int64_t min_element = std::numeric_limits<int64_t>::max(); 
+
+          for(int64_t ele: int_list) {
+            if(ele < min_element) {
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/csrc/jit/register_prim_ops.cpp]: Unfortunately c10::List does not allow me to use it :(
